lib/src/memory/cache.dart

```diff
@@ -0,0 +1,483 @@
+  /// policy is used to choose which way to evict on a write miss.
+  Cache(Logic clk, Logic reset, List<ValidDataPortInterface> writes,
+      List<ValidDataPortInterface> reads,
+      {this.ways = 1,
+      this.lines = 16,
+      this.replacement = PseudoLRUReplacement.new,
+      super.name = 'Cache',
+      super.reserveName,
+      super.reserveDefinitionName,
+      String? definitionName})
+      : dataWidth = (writes.isNotEmpty)
+            ? writes[0].dataWidth
+            : (reads.isNotEmpty)
+                ? reads[0].dataWidth
+                : 0,
+        super(
+            definitionName: definitionName ??
+                'Cache_WP${writes.length}'
+                    '_RP${reads.length}_W${ways}_L$lines') {
+    addInput('clk', clk);
+    addInput('reset', reset);
+    for (var i = 0; i < writes.length; i++) {
+      this.writes.add(writes[i].clone()
+        ..connectIO(this, writes[i],
+            inputTags: {DataPortGroup.control, DataPortGroup.data},
+            uniquify: (original) => 'cwr_${original}_$i'));
+    }
+    for (var i = 0; i < reads.length; i++) {
+      this.reads.add(reads[i].clone()
+        ..connectIO(this, reads[i],
+            inputTags: {DataPortGroup.control},
+            outputTags: {DataPortGroup.data},
+            uniquify: (original) => 'crd_${original}_$i'));
+    }
+    _buildLogic();
+  }
+  @mustBeOverridden
+  void _buildLogic();
+
+  /// Extract the tag from the address.
+  Logic getTag(Logic addr) => addr.getRange(log2Ceil(lines));
+
+  /// Extract the line index from the address.
+  Logic getLine(Logic addr) => addr.slice(log2Ceil(lines) - 1, 0);
+}
+
+/// A multi-ported cache.
+class MultiPortedCache extends Cache {
+  /// Constructs a [Cache] supporting multiple read and write ports.
+  ///
+  /// Defines a set-associativity of [ways] and a depth or number of [lines].
+  /// The total capacity of the cache is [ways]*[lines]. The [replacement]
+  /// policy is used to choose which way to evict on a write miss.
+  MultiPortedCache(super.clk, super.reset, super.writes, super.reads,
+      {super.ways, super.lines, super.replacement});
+
+  @override
+  void _buildLogic() {
+    final numReads = reads.length;
+    final numWrites = writes.length;
+    final lineAddrWidth = log2Ceil(lines);
+    final tagWidth = reads[0].addrWidth - lineAddrWidth;
+
+    final validTagRFMatchWr = _genValidTagRFInterfaces(
+        writes, tagWidth, lineAddrWidth,
+        prefix: 'match_wr');
+    final validTagRFMatchRd = _genValidTagRFInterfaces(
+        reads, tagWidth, lineAddrWidth,
+        prefix: 'match_rd');
+    final validTagRFAlloc = _genValidTagRFInterfaces(
+        writes, tagWidth, lineAddrWidth,
+        prefix: 'alloc');
+
+    // The Tag `RegisterFile`.
+    for (var way = 0; way < ways; way++) {
+      // Combine the read and write match ports for this way.
+      final validTagRFMatch = validTagRFMatchWr[way]
```

Comment:
this is actually modifying the `validTagRFMatchWr`, which from your other comment sounds like it's not your intent?  maybe you want this?
```dart
[
    ...validTagRFMatchWr[way],
    ...validTagRFMatchRd[way]
]
```